Automattic/woocommerce-payments
# Limit enabled payment methods by site currency

Fixes #2418

#### Changes proposed in this Pull Request


# Limiting UPE Payment Methods by Currency

> The goal of this issue is to make sure that the payment types are limited based on the currency the store or customer is currently using. For example, if USD is selected, Giropay should not be offered. We will be getting a thorough list which payment methods accept with currencies from Stripe to implement this.
>
> It should also be noted that until #2391 is implemented, the currency will only be using the store's selected currency. Once #2391 is implemented though, the same check will return the customer's selected currency is multi-currency is[sic.] enabled.

This PR adds a new `currencies` parameter to UPE payment method, which decides which presentment currencies will be supported by a payment method. The parameter can be left empty if there are no currency limits on a payment method. Currently, Giropay and Sofort only support charges in `EUR`, whereas there are no limits on card payments. I am not sure if [cards having no limits is totally accurate](https://stripe.com/docs/currencies#presentment-currencies), since this only appears to be the case for visa and mastercard, but I believe this is an assumption we have been operating under. 

This currency validation is currently only performed when payment or setup intents are created, however, since the use of this function is to limit the payment method present within a mounted UPE, I believe this is the only place where this check is necessary.





#### Testing instructions


1. Please pull these changes, enable UPE, and pull the latest version of the Payments Server (hopefully, you know the drill by now). It will be helpful to also enable multi-currency as well, though you could still test this change by toggling your site's default currency instead of using the client's currency.
2. Add an item to your cart and proceed to checkout. Set currency to USD. 
3. Notice that the only payment method available in the UPE is the card method.
4. Set currency to EUR.
5. Notice that now all UPE payment methods are available in the UPE.
6. Ensure that payment methods available match the currency selected in the other places where UPEs may be present*.
7. Quickly have a peep to see nothing else is broken.

*I noticed that when I was testing this on the Pay for Order page, Giropay and Sofort were removed because for some reason `WC_Subscriptions_Cart::cart_contains_subscription()` returned `true`. I'm not sure why this happened and will update this once I've double-checked this and can present some helpful explanation.




-------------------
